### PR TITLE
[FLINK-33527][autoscaler] Clear all physical states after autoscaler is disabled

### PR DESCRIPTION
**Pull request by 1996fanrui**
## What is the purpose of the change

Currently, we just clear ParallelismOverrides  after autoscaler is disabled.

We should clear CollectedMetrics and ScalingHistory  as well to prevent state leak.

- CollectedMetrics can be cleared directly.
- ScalingHistory can be cleared based on trim logic( VERTEX_SCALING_HISTORY_AGE)

## Brief change log

[FLINK-33527][autoscaler] Clear all physical states after autoscaler is disabled

## Verifying this change

This change added tests and can be verified as follows:

  - *JobAutoScalerImplTest#testAutoscalerDisabled*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changes to the `CustomResourceDescriptors`:  no
  - Core observer or reconciler logic that is regularly executed: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

**Review by 1996fanrui — COMMENTED** [review]
Hey @gyfora  @mxm , this PR fixed some state leaks when autoscaler switches from enabled to disabled. Please help take a look in your free time thanks~

File: flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobAutoScalerImpl.java [hunk 1]
```diff
@@ -115,10 +123,39 @@ public void cleanup(KEY jobKey) {
...
+        }
+
+        Optional<SortedMap<Instant, CollectedMetrics>> collectedMetricsOpt =
+                stateStore.getCollectedMetrics(ctx);
+        if (collectedMetricsOpt.isPresent()) {
+            needFlush = true;
+            stateStore.removeCollectedMetrics(ctx);
+        }
+
+        Optional<Map<JobVertexID, SortedMap<Instant, ScalingSummary>>> scalingHistoryOpt =
```
Comment (gyfora) [inline flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobAutoScalerImpl.java:141]: lets be consistent and use:
```
var scalingHistoryOpt = ...
...
var scalingHistory = ...
```

File: flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobAutoScalerImpl.java [hunk 2]
```diff
@@ -115,10 +123,39 @@ public void cleanup(KEY jobKey) {
...
+            if (scalingHistory.isEmpty()) {
+                // All scaling histories are trimmed.
+                needFlush = true;
+                stateStore.removeScalingHistory(ctx);
+            } else if (!scalingHistoryOpt.get().equals(scalingHistory)) {
+                // Some scaling histories are trimmed.
+                needFlush = true;
+                stateStore.storeScalingHistory(ctx, scalingHistory);
+            }
+        }
```
Comment (gyfora) [inline flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobAutoScalerImpl.java:156]: This looks strange. Why not simply call ` stateStore.removeScalingHistory(ctx);` all the time? Then we wouldn't need to touch the trimming logic.

File: flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobAutoScalerImpl.java (hunk 1 above)
Comment (1996fanrui) [reply flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobAutoScalerImpl.java]: Updated

File: flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobAutoScalerImpl.java (hunk 2 above)
Comment (1996fanrui) [reply flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobAutoScalerImpl.java]: Trimming logic will consider `VERTEX_SCALING_HISTORY_AGE` option, and the default value is `24 hours`.

My thought is: if users enable autoscaler within 24 hours, the `ScalingHistory` can be used for subsequent scaling. WDYT?

Comment (gyfora) [reply flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobAutoScalerImpl.java]: I see, makes sense

**Pull request merged**